### DISCUSSION TITLE
An Example For Adding Testing codes In AppFlowy Editor

### DISCUSSION THREAD

**Discussion by LucasXu0** [AppFlowy-IO/AppFlowy#1107] [Show and tell]
Hi, All.

As the below screenshot shows, there are some files that have less than 70% code coverage. We aim to maintain the code coverage of each code file at 70% and suggest giving preference to files with low code coverage to write your first test.

<img width="1594" src="https://user-images.githubusercontent.com/11863087/191403205-6a9796b0-dc89-4f98-9c46-6b6890642276.png">

> By the way, I implemented a shortcut event by following this [discussion](https://github.com/AppFlowy-IO/AppFlowy/discussions/1100) and wrote my first [test code](https://github.com/AppFlowy-IO/AppFlowy/pull/1099/files#diff-bd88c3a60c2260e2dd2b7280655ac2cc3fde423c2c89493389f8e83851ce59d7).

Here are my steps and a [pull request](https://github.com/AppFlowy-IO/AppFlowy/pull/1099) example.

1. Clone the appflowy project.
2. Open the appflowy_editor project under `frontend/app_flowy/packages/appflowy_editor` in Visual Studio.
3. Create a new dart file called `markdown_syntax_to_styled_text_test.dart` under `test/service/internal_key_event_handlers`.
4. Mock to input `**AppFlowy*`.
5. Mock to input `shift+asterisk` to trigger the `doubleAsterisksToBold` shortcut .
6. Expect the plain text `**AppFlowy**` to have converted to **AppFlowy**.
7. Click the `run` button in Visual Studio Code and check the result.
8. Done!

**Reference**
[Codecov For AppFlowy Editor](https://app.codecov.io/gh/AppFlowy-IO/AppFlowy/tree/main/frontend/app_flowy/packages/appflowy_editor/lib/src)
[Testing](https://github.com/AppFlowy-IO/AppFlowy/blob/main/frontend/app_flowy/packages/appflowy_editor/documentation/testing.md)

**Comment by annieappflowy** [comment]
https://www.browserstack.com/guide/code-coverage-vs-test-coverage
wonder code coverage v.s. test coverage? I just learnt the differences!
